### PR TITLE
Headers in SerializationContext

### PR DESCRIPTION
**Pull request by ldom**

### CODE REVIEW

**Comment by edenhill** [comment]
It'd be good with a simple test to make sure they're passed correctly.

And there's also a bunch of tests that instantiate a SerializationContext() without passing this argument, which are likely to fail now. Perhaps make headers optional (None) or update those tests, or both.

**Comment by ldom** [comment]
I'll do both

File: tests/schema_registry/test_avro_serializer.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def test_avro_serializer_record_subject_name_strategy(load_avsc):
                                      conf={'subject.name.strategy':
                                            record_subject_name_strategy})
 
-    ctx = SerializationContext('test_subj', MessageField.VALUE)
+    ctx = SerializationContext('test_subj', MessageField.VALUE, [])
```
Comment (edenhill) [inline tests/schema_registry/test_avro_serializer.py:114]: From a test perspective we yield better coverage if some of these use the default value (not specified), and some use an empty list

**Review by mhowlett — APPROVED** [review]

**Review by edenhill — CHANGES_REQUESTED** [review]
There needs to be at least one test that verifies that passed headers are in-fact available on the ctx, preferably a negative test too (no headers in, no headers on ctx)

**Comment by ldom** [comment]
I've added a few tests. Some are an overkill but there is never too many tests.

File: src/confluent_kafka/serialization/__init__.py [hunk 2]
```diff
@@ -56,10 +56,13 @@ class SerializationContext(object):
         field (MessageField): Describes what part of the message is
             being serialized.
 
+        headers (List of header tuples): Message headers
```
Comment (edenhill) [inline src/confluent_kafka/serialization/__init__.py:59]: Should say that it may be None

File: tests/schema_registry/test_avro_serializer.py [hunk 3]
```diff
@@ -176,3 +176,66 @@ def test_avro_serializer_subject_name_strategy_default(load_avsc):
...
+    test_serializer = AvroSerializer(test_client, 'int',
+                                     conf={'subject.name.strategy':
+                                           record_subject_name_strategy})
+
+    headers = [('k1', 'v1'),]
+
+    ctx = SerializationContext('test_subj', MessageField.VALUE, headers)
+    assert ('k1', 'v1') in ctx.headers
+    headers.extend([('k2', 'v2'),])
+    assert ('k2', 'v2') in ctx.headers
```
Comment (edenhill) [inline tests/schema_registry/test_avro_serializer.py:240]: nit: This is a side-effect of not copying the headers, which may not even be desirable, so not sure this is something we should test for.

File: tests/schema_registry/test_avro_serializer.py [hunk 4]
```diff
@@ -176,3 +176,66 @@ def test_avro_serializer_subject_name_strategy_default(load_avsc):
...
+    assert len(headers) == 2
```
Comment (edenhill) [inline tests/schema_registry/test_avro_serializer.py:241]: I prefer as exact as possible matching in tests, and order matters for the headers, so suggest doing an exact match between ctx.headers and what is expected in there.

**Review by edenhill — CHANGES_REQUESTED** [review]
Good stuff!

File: tests/schema_registry/test_avro_serializer.py (hunk 3 above)
Comment (ldom) [reply tests/schema_registry/test_avro_serializer.py:240]: removed

Comment (ldom) [reply tests/schema_registry/test_avro_serializer.py:240]: not really useful anyway

File: src/confluent_kafka/serialization/__init__.py [hunk 5]
```diff
@@ -56,10 +56,13 @@ class SerializationContext(object):
         field (MessageField): Describes what part of the message is
             being serialized.
 
+        headers (optional, defaults to None, List of header tuples): Message headers
```
Comment (edenhill) [inline src/confluent_kafka/serialization/__init__.py:59]: I think the paren may only hold the type, so reframe as:
```python
  headers (list): List of message header tuples. Defaults to None.
```

**Comment by edenhill** [comment]
Looks good!

Just need to sort out the flake8 errors:
https://travis-ci.org/github/confluentinc/confluent-kafka-python/jobs/769668688#L803

(run flake8 locally to find all style issues)
